README.md
# Shell Scripts For WordPress Theme & WordPress Plugin Developers - WP-Cli-Localhost v1.1.0

Created for WordPress Theme and WordPress Plugin developers. WP-Cli-Localhost is a collection of shell scripts that interface with wp-cli to allow for quick and easy management of locally hosted WordPress websites.

* View The [Shell Scripts For WordPress Developers Documentation](https://chriswinters.github.io/wp-cli-localhost/) On Github Pages

1) Run With Alias Shell Function Command "domain"
	* Works From Any Location
2) Rapid Setup of Localhost Domain & WordPress
	* Automatically Updates Apache and /etc/hosts
	* Cleans WordPress Setup, No Content
	* Developer & Debug Plugins Installed
	* Advanced Debug Backtrace Configuration
3) Rapid Removal of Localhost Domain & WordPress 
	* WordPress, Database, Apache and Hosts Configurations
4) Automatically Update Localhost WordPress Installs
	* Automatically Loops Through Locally Created Domains
		* Or Directly Target A Domain
	* Automatically Updates WordPress, Themes, and Plugins
5) Import WordPress Unit Test Data To Any Website

## Requirements

1) Requires a LAMP Stack (Linux, Apache, MySQL, PHP)
	* Built and tested on Linux Mint using Bash 4.3.11(1)-release
2) Requires [WP-CLI](https://wp-cli.org/)
	* Built using version 2.1.0
3) Configuration

[Git](https://git-scm.com/downloads) is optional but recommended.


## Download / Clone

[Download](https://github.com/ChrisWinters/wp-cli-localhost/releases) or [clone](https://github.com/ChrisWinters/wp-cli-localhost.git) WP-Cli-Localhost to an accessible location for your user (not /Private), something like ``` ~/ ``` or the location of localhost domains ``` /var/www/html ```.

Then open the ``` wp-cli-localhost ``` directory.


## Installation

Set the proper file permission on the install.sh file and run the script.

```
cd wp-cli-localhost/
chmod +x install.sh
./install.sh
source ~/.bashrc
```

This will add the 'domain' shell function alias to your .bashrc file, and set the file permissions for all other shell scripts. As well, the files ``` domains.log ``` & ``` config.sh ``` were created, which have been added to the .gitignore file.

**Errors:** If you are still receiving permission errors, first try resetting the file permissions on the install.sh file ``` chmod 644 install.sh; chmod +x install.sh; ```. If you are still receiving errors, set the proper permissions on the /wp-cli-localhost/ directory: ``` cd ../; chmod 755 wp-cli-localhost; ```


## Configuration

Within the ``` wp-cli-localhost ``` directory open the user ``` config.sh ``` file in an editor. The settings within the user config.sh override the settings within domain-config.sh file.

The default config.sh includes settings typically used for WordPress installs.

01) DB_USER 			Update If Different
02) DB_PASS 			Set Proper Password
03) DB_HOST 			Update If Different
04) ADMIN_USER 			WordPress Admin Username
05) ADMIN_PASS 			WordPress Admin Password
06) ADMIN_EMAIL 		Set If Your Localhost Has Mail
07) GMT_OFFSET 			Your Local GMT Offset
08) CATEGORY_BASE 		WordPress Category Base	
09) CATEGORY_NAME 		WordPress Default Category Name
10) CATEGORY_SLUG 		WordPress Default Category Slug
11) TAG_BASE 			WordPress Tag Base


## Commands

Type 'domain' for a quick usage reference.

**Add or Remove A Domain and WordPress**
* Usage: domain add|del domain-name.extension
	* Example: domain add testdomain.localhost
	* Example: domain del testdomain.localhost

**Update WordPress, Themes, and Plugins**
* Usage: domain update all|domain-name.extension
	* Example: domain update all
	* Example: domain update testdomain.localhost


## Upgrading

Git users simply need to ``` git pull ``` the changes, otherwise [download](https://github.com/ChrisWinters/wp-cli-localhost/releases) the latest release and replace the files. The ``` config.sh ``` and ``` domains.log ``` files are unique to your install and need to remain in the active wp-cli-localhost location.

Check the [CHANGELOG.md](https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/CHANGELOG.md) for release notes.


## Removing

1) Delete the wp-cli-localhost directory
2) Optional: The domain() function was added to ``` ~/.bashrc ```
	* This function will not error if wp-cli-localhost is missing
	* To remove the domain function, edit your ``` ~/.bashrc ``` as root, delete the entire domain() function and the export line below the function. Once done type ``` source ~/.bashrc ```.


## Donations

If WP-Cli-Localhost saved you some time, helped you with your development needs, or if you would like to support my efforts, then I would gladly and greatly appreciate a donated cup of coffee. Thank you in advance for your support!

<a href="https://www.buymeacoffee.com/TavXZIxkm" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>


## To Do List

A wish list of things that may or may not get completed.

* Move install.sh to Makefile
* Move domains.log to NoSQL
* Add check for backtrace.php to make sure wp-config.php has permission to include from backtrace location
* Add version check command that responds with the current version and release version with update notice
* Port to various Operating Systems, starting with Windows 10


## Credits

* [Enhanced Debugging](https://gist.github.com/jrfnl/5925642)
* [Theme Unit Test Data](https://github.com/WPTRT/theme-unit-test)


## MIT License

[License](https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE)

Copyright 2019 Chris Winters https://github.com/ChrisWinters/

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
<a name="1.1.0"></a>
# 1.1.0 - 2019-02-07 - Feature Release

* Added 'import' command to domain() | Imports Theme Unit Test Data
* Moved domain() to .domain-bashrc & updated install.sh to use new source for .bashrc
	* This will allow for .domain-bashrc to be re-generated if domain() updates

<a name="1.0.0"></a>
# 1.0.0 - 2019-02-05 - Public Release

* Initial Github Release


backtrace.php
<?php
/**
 * == About this Gist ==
 * 
 * Code to add to wp-config.php to enhance information available for debugging.
 * You would typically add this code below the database, language and salt settings
 * 
 * Oh.. and *do* make sure you change the path to the log file to a proper file path on your server (make sure it exists).
 * 
 * Set WP_DEBUG_DISPLAY to false if you don't want errors displayed on the screen.
 * Independently of the WP_DEBUG settings, PHP errors, warnings and notices should now appear in your error.log file.
 * 
 * == DO == test whether it's all working by using the code at the end of the gist.
 * 
 * If error logging is not working, try:
 * - Check if the file was created at all, if not, upload an empty (text) file named error.log and try again.
 * - Fiddle around with the CHMOD settings for the error.log file to see what permissions are needed for your server setup.
 * - If you put the error.log outside of the web root: this will not work in all server setups. You may need to move the directory & file to be within the web root.
 *
 * Source: https://gist.github.com/jrfnl/5925642
 */

/**
 * For developers: WordPress debugging mode.
 *
 * Change this to true to enable the display of notices during development.
 * It is strongly recommended that plugin and theme developers use WP_DEBUG
 * in their development environments.
 */

/**
 * Turn on WP error reporting.
 */
define( 'WP_DEBUG', true );



/**
 * [Optional] Load the development (non-minified) versions of all standard WP scripts and CSS, and disables compression and concatenation.
 *
 * This setting is often also respected by themes and plugins, but no guarantee!
 */
// define( 'SCRIPT_DEBUG', true );


/**
 * [Optional] Save queries for analysis
 * The SAVEQUERIES definition saves the database queries to an array and that array
 * can be displayed to help analyze those queries. The information saves each query,
 * what function called it, and how long that query took to execute.
 *
 * NOTE: This will have a performance impact on your site, so make sure to turn this
 * off when you aren't debugging.
 *
 * To get access to the information, you can add the following snippet to the footer of your theme:
 *
 * <?php
 *	if ( current_user_can( 'administrator' ) ) {
 * 		global $wpdb;
 * 		echo "<pre>";
 * 		print_r($wpdb->queries);
 * 		echo "</pre>";
 *	}
 * ?>
 */
// define('SAVEQUERIES', true);



/**
 * Turn on error logging and show errors on-screen if in debugging mode.
 */
@error_reporting( -1 ); // everything, including E_STRICT and other newly introduced error levels.
@ini_set( 'log_errors', true );
@ini_set( 'log_errors_max_len', '0' );

/**
 * Change the path to one on your webserver, the directory does not have to be in the web root
 * Don't forget to CHMOD this dir+file and add an .htaccess file denying access to all
 * For an example .htaccess file, see https://gist.github.com/jrfnl/5953256
 */
//@ini_set( 'error_log', '/path/to/writable/file/logs/error.log' );

if ( WP_DEBUG !== true ) {
	@ini_set( 'display_errors', false ); // Don't show errors on screen
}
else {
	@ini_set( 'display_errors', true ); // Show errors on screen
	@ini_set( 'html_errors', true );
	@ini_set( 'docref_root', 'http://php.net/manual/' );
	@ini_set( 'docref_ext', '.php' );
	if ( ! extension_loaded( 'xdebug' ) ) {
		@ini_set( 'error_prepend_string', '<span style="color: #ff0000; background-color: transparent;">' );
		@ini_set( 'error_append_string', '</span>' );
	}
}


/**
 * Adds a backtrace to PHP errors.
 * 
 * Copied from: https://gist.github.com/625769
 * Forked from: http://stackoverflow.com/questions/1159216/how-can-i-get-php-to-produce-a-backtrace-upon-errors/1159235#1159235
 * Adjusted by jrfnl.
 */
if ( ! function_exists( 'jrf_process_error_backtrace' ) ) {
	function jrf_process_error_backtrace( $errno, $errstr, $errfile, $errline ) {
		/*
		 * Only show errors which are within the scope of the current error_reporting() setting.
		 * As WP (and plugins/themes) sometimes change the error level, you may want to comment
		 * this out if you want to be sure to see all errors.
		 */
		if( ! ( error_reporting() & $errno ) ) {
			return;
		}

		/**
		 * Make sure all error levels are present. PHP 5.2 compatibility.
		 */
		if ( ! defined( 'E_DEPRECATED' ) ) {
			define( 'E_DEPRECATED', 8192 );
		}
		if ( ! defined( 'E_USER_DEPRECATED' ) )	{
			define( 'E_USER_DEPRECATED', 16384 );
		}
		
	
		switch ( $errno ) {
			case E_WARNING	    :
			case E_USER_WARNING :
			case E_STRICT	       :
			case E_NOTICE	       :
			case E_USER_NOTICE     :
			case E_DEPRECATED      :
			case E_USER_DEPRECATED :
				$type  = 'warning';
				$fatal = false;
				break;
			default			 :
				$type  = 'fatal error';
				$fatal = true;
				break;
		}
		$trace = debug_backtrace();
		array_shift( $trace );
		if ( php_sapi_name() == 'cli' && ini_get( 'display_errors' ) ) {
			echo 'Backtrace from ' . $type . ' \'' . $errstr . '\' at ' . $errfile . ' ' . $errline . ':' . "\n";
			foreach ( $trace as $item ) {
				echo '  ' . ( isset( $item['file'] ) ? $item['file'] : '<unknown file>' ) . ' ' . ( isset( $item['line'] ) ? $item['line'] : '<unknown line>' ) . ' calling ' . $item['function'] . '()' . "\n";
			}
		}
		elseif ( ini_get( 'display_errors' ) ) {
			echo '<p class="error_backtrace">' . "\n";
			echo '  Backtrace from ' . $type . ' \'' . $errstr . '\' at ' . $errfile . ' ' . $errline . ':' . "\n";
			echo '</p>' . "\n";
			echo '<ol class="error_backtrace_list">' . "\n";
			foreach ( $trace as $item ) {
				echo '	<li>' . ( isset( $item['file'] ) ? $item['file'] : '<unknown file>' ) . ' ' . ( isset( $item['line'] ) ? $item['line'] : '<unknown line>' ) . ' calling ' . $item['function'] . '()</li>' . "\n";
			}
			echo '</ol>' . "\n";
		}
		if (  ini_get( 'log_errors' ) ) {
			$items = array();
			foreach ( $trace as $item ) {
				$items[] = ( isset( $item['file'] ) ? $item['file'] : '<unknown file>' ) . ' ' . ( isset( $item['line'] ) ? $item['line'] : '<unknown line>' ) . ' calling ' . $item['function'] . '()';
			}
			$message = 'Backtrace from ' . $type . ' \'' . $errstr . '\' at ' . $errfile . ' ' . $errline . ': ' . trim( join( ' | ', $items ) );
			error_log( $message );
		}
	
		flush();

		if ( $fatal ) {
			exit( 1 );
		}

		return false; // Make sure it plays nice with other error handlers (remove if no other error handlers are set).
	}

	set_error_handler( 'jrf_process_error_backtrace' );
}


/**
 * Now test whether it all works by uncommenting the below line
 *
 * If all is well:
 * - With WP_DEBUG set to true: You should see a red error notice on your screen 
 * - Independently of the WP_DEBUG setting, the below 'error'-message should have been written to your log file. *Do* check whether it has been....
 */
//trigger_error( 'Testing 1..2..3.. Debugging code is working!', E_USER_NOTICE );


domain-config.sh
#! /bin/bash
#
# Default Domain Configuration File
#
# DO NOT UPDATE THIS FILE : Use config.sh
# DO NOT UPDATE THIS FILE : Use config.sh
# DO NOT UPDATE THIS FILE : Use config.sh
#
# @version 1.0.0
#
# @author Chris Winters https://github.com/ChrisWinters
#
# MIT License
# https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE

###################################
##                               ##
##     DO NOT EDIT THIS FILE     ##
##      use config.sh file!      ##
##                               ##
###################################

# Domain Name Parameter
# $2 From Shell Function "domain"
DOMAIN=$1;

# Path To Local Domains
SITE_PATH="/var/www/html";

# Base Localhost URL (http or https?)
LOCAL_DOMAIN_URL="http://${DOMAIN}";

# Local Hosts IP
HOSTS_IP="127.0.0.1";

# Database Login Details
# mysql_config_editor set --login-path=local --host=localhost --user=root --password
DB_USER="";
DB_PASS="";
DB_HOST="";

# WordPress Super User
ADMIN_USER="";
ADMIN_PASS="";
ADMIN_EMAIL="";

# WordPress GMT Offset
GMT_OFFSET="";

# WordPress Category Base
CATEGORY_BASE="";

# Default WordPress Category Name & Slug
CATEGORY_NAME="";
CATEGORY_SLUG="";

# WordPress Tag Base
TAG_BASE="";

# Default Permalink Structure
PERMALINK="/%postname%/";

# WordPress Dev Error Log
ERROR_LOG="${SITE_PATH}/error.log";

# Path WP-Cli-Localhost
WP_CLI_LOCALHOST_PATH=$(pwd);

# Localhost Domains List
DOMAINS_LIST="${WP_CLI_LOCALHOST_PATH}/domains.log"

# Full Path To Localhost Domain
LOCAL_DOMAIN_PATH="${SITE_PATH}/${DOMAIN}";

# Host Entry Record
HOSTS_ENTRY="${HOSTS_IP} ${DOMAIN}";

# Database Name
DB_NAME=$(echo ${DOMAIN} | sed -e 's/-/_/g' | cut -f 1 -d '.');

# Include User config.sh File
if [[ -f ${WP_CLI_LOCALHOST_PATH}/config.sh ]]; then
	source ${WP_CLI_LOCALHOST_PATH}/config.sh;
fi


domain-import.sh
#! /bin/bash
#
# Download / Install Theme Unit Test Data
#
# @cmd domain import domainname.extension
#
# @version 1.0.0
#
# @author Chris Winters https://github.com/ChrisWinters
#
# MIT License
# https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE

# Source Config File
source domain-config.sh


# Download Unit Test Data
function get_xml_file() {
	if [[ ! -f themeunittestdata.wordpress.xml ]]; then
		wget https://raw.githubusercontent.com/WPTRT/theme-unit-test/master/themeunittestdata.wordpress.xml
		printf "===> Theme Unit Test Data XML File Downloaded\n"
	fi
}


# Active WordPress Importer Plugin
function activate_importer () {
	wp plugin install wordpress-importer --activate --path=${LOCAL_DOMAIN_PATH}
	printf "===> WordPress Importer Activated\n"
}


# Run WP Import
function wp_import () {
	printf "===> Starting Import, This Will Take A Few Minutes...\n\n"
	wp import themeunittestdata.wordpress.xml --authors=create --path=${LOCAL_DOMAIN_PATH}
	printf "\n===> Import Completed\n"
}


# Run Remove Domain
function import() {
	# Required
    if [[ $1 == "" ]]; then
	    echo $0: "=domain=> usage example: domain import domainname.extension"
	    return 1
    fi

	# Start Display
	clear

	printf "\n\n"

	read -p "== Press [y] to IMPORT Theme Unit Test Data for ${LOCAL_DOMAIN_URL} " -n 1 -r

	printf "\n"

	if [[ $REPLY =~ ^[Yy]$ ]]; then
		printf "=domain=> Working on ${DOMAIN}\n"
		sleep 1

		get_xml_file
		activate_importer
		wp_import

		printf "=domain=> Ctrl+Click To Open: ${LOCAL_DOMAIN_URL}\n\n"
	fi
}

import $@


domain-remove.sh
#! /bin/bash
#
# Automatic Localhost Domain & WordPress Removal
#
# @cmd domain remove domainname.extension
#
# @version 1.0.0
#
# @author Chris Winters https://github.com/ChrisWinters
#
# MIT License
# https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE

# Source Config File
source domain-config.sh


# Remove lcoal domain directory
function remove_local_domain(){
	if [[ -d "${SITE_PATH}/${DOMAIN}" ]]; then
		rm -rf ${SITE_PATH}/${DOMAIN}/
		printf "===> Removed WordPress\n"
	fi
}


# Remove database
function remove_database(){
	mysql --login-path=local -e "DROP DATABASE $DB_NAME"
	printf "===> Removed Database $DB_NAME\n"
}


# Update /etc/hosts
function update_hosts_file() {
	printf "===> Updating Hosts Entry; You May Need To Enter Your Password\n"

	sudo sh -c "grep -Fxv \"${HOSTS_ENTRY}\" /etc/hosts > /etc/hosts.temp"

	if [[ -f "/etc/hosts.temp" ]]; then
		sudo sh -c "cp /etc/hosts.temp /etc/hosts"
		sudo sh -c "rm /etc/hosts.temp"

		printf "===> Updated /etc/hosts\n"
	fi
}


# Remove domain from Apache
function update_apache(){
	printf "===> Removing Domain From Apache; You May Need To Enter Your Password\n"

	sudo sh -c "unlink /etc/apache2/sites-enabled/${DOMAIN}.conf"
	sudo sh -c "rm /etc/apache2/sites-available/${DOMAIN}.conf"

	printf "===> Restarting Apache\n"

	sudo sh -c "apache2ctl -k graceful"
}


# Update Domain Log
function update_log() {
	grep -Fxv "${DOMAIN}" ${DOMAINS_LIST} > ${DOMAINS_LIST}.temp

	if [[ -f "${DOMAINS_LIST}.temp" ]]; then
		cp ${DOMAINS_LIST}.temp ${DOMAINS_LIST}
		rm ${DOMAINS_LIST}.temp

		printf "===> Removed Domain From Log\n"
	fi
}


# Run Remove Domain
function remove_domain(){
	# Required
    if [[ $1 == "" ]]; then
	    echo $0: "=domain=> usage example: domain remove domainname.extension"
	    return 1
    fi

	# Start Display
	clear

	printf "\n\n"

	read -p "== Press [y] to REMOVE ${LOCAL_DOMAIN_URL} " -n 1 -r

	printf "\n\n"

	if [[ $REPLY =~ ^[Yy]$ ]]; then
		printf "=domain=> Working on ${DOMAIN}\n"
		sleep 1

		remove_local_domain
		remove_database
		update_hosts_file
		update_apache
		update_log

		printf "=domain=> Domain ${DOMAIN} Removed\n\n"
	fi
}

remove_domain $@


domain-update.sh
#! /bin/bash
#
# Update WordPress, Plugins & Themes
#
# @cmd domain update all
# @cmd domain update domainname.extension
#
# @version 1.0.0
#
# @author Chris Winters https://github.com/ChrisWinters
#
# MIT License
# https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE

# Run Updater
update() {
	# Required
    if [[ $1 == "" ]]; then
	    echo $0: "=domain=> usage example: domain update all"
	    echo $0: "                         domain update domainname.extension"
	    return 1
    fi

	# Source Config File
	source domain-config.sh

	# Required
    if [[ $1 != "all" && ! -d ${SITE_PATH}/${1} ]]; then
	    echo $0: "=domain=> usage example: domain update all"
	    echo $0: "                         domain update domainname.extension"
	    return 1
    fi

    # Update All Domains
    if [[ $1 == "all" ]]; then
		printf "\n=domain=> Updating All Domains\n"
		sleep 1

		while read DOMAIN_NAME; do
			printf "==> Updating: ${DOMAIN_NAME}\n"
			wp core update --path=${SITE_PATH}/${DOMAIN_NAME}
			wp plugin update --path=${SITE_PATH}/${DOMAIN_NAME} --all
			wp theme update --path=${SITE_PATH}/${DOMAIN_NAME} --all
		done < ${WP_CLI_LOCALHOST_PATH}/domains.log

		printf "\n=domain=> Update Completed\n\n"
	fi

	# Update Single Domain
	if [[ $1 != "" && -d ${SITE_PATH}/${1} ]]; then
		printf "=domain=> Updating ${1}\n"
		sleep 1

		wp core update --path=${SITE_PATH}/${1}
		wp plugin update --path=${SITE_PATH}/${1} --all
		wp theme update --path=${SITE_PATH}/${1} --all

		printf "\n=domain=> Update Completed\n\n"
	fi
}

update $@


install.sh
#! /bin/bash
#
# Install WP-Cli-Localhost
#
# @cmd ./install.sh
#
# @version 1.0.0
#
# @author Chris Winters https://github.com/ChrisWinters
#
# MIT License
# https://raw.githubusercontent.com/ChrisWinters/wp-cli-localhost/master/LICENSE

# Source Config File
source domain-config.sh


# Create Local .domain-bashrc File
create_domain_bashrc() {
	# Remove Old File
	if [[ -f ${WP_CLI_LOCALHOST_PATH}/.domain-bashrc ]]; then
		rm ${WP_CLI_LOCALHOST_PATH}/.domain-bashrc
	fi

	# Generate File
	echo -e  "
# Domain Shell Function
domain() {
    # Two Args Required
    if [[ \$# -ne 2 ]]; then
        echo \$0: \"example ==> domain install|remove|import|update domainname.extension\"
        return 1
    fi

    DIR=\$(pwd);

    # Install Domain
    if [[ \$1 == \"install\" ]] && [[ \$2 != \"\" ]]; then
        cd ${WP_CLI_LOCALHOST_PATH}
        ./domain-install.sh \$2
    fi

    # Remove Domain
    if [[ \$1 == \"remove\" ]] && [[ \$2 != \"\" ]]; then
        cd ${WP_CLI_LOCALHOST_PATH}
        ./domain-remove.sh \$2
    fi

    # Import Test Data
    if [[ \$1 == \"import\" ]] && [[ \$2 != \"\" ]]; then
        cd ${WP_CLI_LOCALHOST_PATH}
        ./domain-import.sh \$2
    fi

    #Update Domain(s)
    if [[ \$1 == \"update\" ]]; then
        cd ${WP_CLI_LOCALHOST_PATH}

        # Update All Domains
        if [[ \$2 == \"all\" ]]; then
            ./domain-update.sh all
        fi

        # Update A Domain
        if [[ \$2 != \"all\" && -d \"\${SITE_PATH}/${2}\" ]]; then
            ./domain-update.sh \$2
        fi
    fi

	cd \${DIR};
}

# Export Alias
export -f domain \$@
" >> ${WP_CLI_LOCALHOST_PATH}/.domain-bashrc
}


# Create Empty domains.log File
create_domain_log() {
	if [[ ! -f ${DOMAINS_LIST} ]]; then
		echo "" > ${DOMAINS_LIST}
	fi
}


# Create User config.sh File
create_config_sh() {
	if [[ ! -f ${WP_CLI_LOCALHOST_PATH}/config.sh ]]; then
		touch ${WP_CLI_LOCALHOST_PATH}/config.sh
		cat > ${WP_CLI_LOCALHOST_PATH}/config.sh <<EOF
#! /bin/bash
#
# User Configuration File | This File Was Generated When WP-Cli-Localhost Was Installed
# Override Variables From domain-config.sh
#

# Database Login Details
DB_USER="root";
DB_PASS="PASSWORD";
DB_HOST="localhost";

# WordPress Super User
ADMIN_USER="WORDPRESS";
ADMIN_PASS="WORDPRESS";
ADMIN_EMAIL="nobody@domain.com";

# WordPress GMT Offset
GMT_OFFSET="-4";

# WordPress Category Base
CATEGORY_BASE="/topics";

# Default WordPress Category Name & Slug
CATEGORY_NAME="Updates";
CATEGORY_SLUG="updates";

# WordPress Tag Base
TAG_BASE="/tags";
EOF
	fi
}


# Update File Permissions
update_permissions() {
	chmod 755 ${WP_CLI_LOCALHOST_PATH}
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/backtrace.php
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/config.sh; chmod +x ${WP_CLI_LOCALHOST_PATH}/config.sh;
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/domain-config.sh; chmod +x ${WP_CLI_LOCALHOST_PATH}/domain-config.sh;
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/domain-install.sh; chmod +x ${WP_CLI_LOCALHOST_PATH}/domain-install.sh;
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/domain-remove.sh; chmod +x ${WP_CLI_LOCALHOST_PATH}/domain-remove.sh;
	chmod 644 ${WP_CLI_LOCALHOST_PATH}/domain-update.sh; chmod +x ${WP_CLI_LOCALHOST_PATH}/domain-update.sh;
}


# Update .bashrc if domain function not set
update_bashrc() {
	if [[ -n "$(type -t domain)" ]] && [[ "$(type -t domain)" = function ]]; then
		printf "==> WP-Cli-Localhost Already Installed\n"
		printf "===> To Configure, open: config.sh\n\n"
		printf "      Commands To Use:\n"
		printf "        Install: domain install domainname.extension\n"
		printf "        Remove : domain remove domainname.extension\n"
		printf "        Import : domain import domainname.extension\n"
		printf "        Update : domain update domainname.extension\n"
		printf "        Update : domain update all\n\n"
	else
		# Source in .bashrc
		echo -e  "
source ${WP_CLI_LOCALHOST_PATH}/.domain-bashrc
" >> ~/.bashrc

		source ~/.bashrc

		if [[ -z "$PS1" ]] ; then
			printf "\n\n"
		    printf "====> Almost Finished, type: source $HOME/.bashrc\n"
		    printf "====> To Configure, open: config.sh\n\n"
			printf "      Then Use The Commands Below:\n"
			printf "        Install: domain install domainname.extension\n"
			printf "        Remove : domain remove domainname.extension\n"
			printf "        Import : domain import domainname.extension\n"
			printf "        Update : domain update domainname.extension\n"
			printf "        Update : domain update all\n\n"
		    printf "====> Almost Finished, type: source $HOME/.bashrc\n"
		    printf "====> To Configure, open: config.sh\n\n"
		else
			source $HOME/.bashrc
			printf "==> WP-Cli-Localhost Installed\n\n"
		    printf "====> To Configure, open: config.sh\n\n"
			printf "      Commands To Use:\n"
			printf "        Install: domain install domainname.extension\n"
			printf "        Remove : domain remove domainname.extension\n"
			printf "        Import : domain import domainname.extension\n"
			printf "        Update : domain update domainname.extension\n"
			printf "        Update : domain update all\n\n"
		    printf "====> To Configure, open: config.sh\n\n"
		fi
	fi
}


# Run Install
function install() {
	# Start Display
	clear

	printf "\n\n"

	read -p "== Press [y] to setup WP-Cli-Localhost " -n 1 -r

	printf "\n\n"

	if [[ $REPLY =~ ^[Yy]$ ]]; then

		create_domain_bashrc
		create_domain_log
		create_config_sh
		update_permissions
		update_bashrc

	fi
}

install
